rs/replicated_state/src/canister_state/queues/tests.rs

```diff
@@ -791,6 +941,260 @@ fn test_skip_input_round_robin() {
+    assert_matches!(queues.pop_input().unwrap(), CanisterMessage::Request(request) if request.sender == canister_test_id(2));
+
+    assert_matches!(queues.peek_input().unwrap(), CanisterMessage::Request(request) if request.sender == canister_test_id(5));
+    assert_matches!(queues.pop_input().unwrap(), CanisterMessage::Request(request) if request.sender == canister_test_id(5));
+
+    assert_eq!(None, queues.peek_input());
+    assert_eq!(None, queues.pop_input());
+
+    assert!(queues.pool.len() == 0);
+}
+
+#[test]
+fn test_peek_input_with_gced_queue_in_input_schedule() {
+    let mut queues = fixture_with_empty_queues_in_input_schedules();
+
+    // Garbage collect the empty queue paurs.
+    queues.garbage_collect();
+    // Only 2 queue pairs should be left.
+    assert_eq!(2, queues.canister_queues.len());
+
+    assert_matches!(queues.peek_input().unwrap(), CanisterMessage::Request(request) if request.sender == canister_test_id(2));
+    assert_matches!(queues.pop_input().unwrap(), CanisterMessage::Request(request) if request.sender == canister_test_id(2));
+
+    assert_matches!(queues.peek_input().unwrap(), CanisterMessage::Request(request) if request.sender == canister_test_id(5));
+    assert_matches!(queues.pop_input().unwrap(), CanisterMessage::Request(request) if request.sender == canister_test_id(5));
+
+    assert_eq!(None, queues.peek_input());
+    assert_eq!(None, queues.pop_input());
+
+    assert!(queues.pool.len() == 0);
+}
+
+#[test]
+fn roundtrip_encode_empty_queue_in_input_schedule() {
+    let queues = fixture_with_empty_queues_in_input_schedules();
+
+    let encoded: pb_queues::CanisterQueues = (&queues).into();
+    let decoded = (encoded, &StrictMetrics as &dyn CheckpointLoadingMetrics)
+        .try_into()
+        .unwrap();
+
+    assert_eq!(queues, decoded);
+}
+
+#[test]
+fn roundtrip_encode_gced_queue_in_input_schedule() {
+    let mut queues = fixture_with_empty_queues_in_input_schedules();
+
+    // Garbage collect the empty queue paurs.
+    queues.garbage_collect();
+    // Only 2 queue pairs should be left.
+    assert_eq!(2, queues.canister_queues.len());
+
+    let encoded: pb_queues::CanisterQueues = (&queues).into();
+    let decoded = (encoded, &StrictMetrics as &dyn CheckpointLoadingMetrics)
+        .try_into()
+        .unwrap();
+
+    assert_eq!(queues, decoded);
+}
+
+#[test]
+fn test_push_into_empty_queue_in_input_schedule() {
+    let other_1 = canister_test_id(1);
+    let other_2 = canister_test_id(2);
+
+    let mut queues = CanisterQueuesMultiFixture::new();
+
+    // 1 local and 1 remote input queue holding best-effort requests.
+    queues
+        .push_input_request_with_deadline(other_1, coarse_time(1), LocalSubnet)
+        .unwrap();
+    queues
+        .push_input_request_with_deadline(other_2, coarse_time(1), RemoteSubnet)
+        .unwrap();
+
+    // Time out all messages.
+    queues.time_out_all_messages_with_deadlines();
```

Comment:
I wouldn't. The only time we actually need to enforce a hard invariant is when mutating the underlying data. At all other times, it is merely a sanity check, in case we introduce new code that both breaks the invariant and doesn't check for it.

Also, if we made it into a soft invariant, it would be a total mess: if we had duplicate schedule entries, as soon as we removed the first duplicate, we'd also remove it from `input_schedule_canisters` and allow another duplicate to be created. Worse, if some canister ID made into `input_schedule_canisters` without it also being scheduled in one of the queues, it would never get scheduled again. Finally, since we populate `input_schedule_canisters` from the contents of the input schedules during checkpoint loading, if its contents became inconsistent with those of the schedules, then by definition they would also diverge between replicas that continued execution across a checkpoint and a replica that started from the same checkpoint.

Reply:
Ok. 

For the first two arguments regarding inconsistencies that may arise I'm not quite convinced that crashing a subnet is really better than going ahead with a critical error. It is not great if a canister never gets scheduled again, but it would seem that nothing catastrophic would break and it seems to be fixable with a hotfix or with the next replica release. 

I guess the last argument about divergence is a good one: even if we'd go ahead when we observe that an invariant is broken we'd at most make it to the next checkpoint or a couple rounds after that. So we don't win anything if we don't strictly check here as we couldn't go ahead much further with the non-strict check anyways.

Reply:
That would amount to working around divergence caused by bugs internal to `CanisterQueues` by adding yet more `CanisterQueues` code; plus an external source of divergence (a redundant collection that is loaded from a checkpoint and cannot be adjusted if any divergence is detected).